### PR TITLE
[feature/licensing] Licensing support

### PR DESCRIPTION
**Pull request by felix-schwarz**
## Description
This pull request adds a completely modular licensing subsystem that can be used to handle and enforce licensing information from various sources, incl. the App Store and ownCloud-provided sources.

## Related Issue
#555 

## Screenshots (if appropriate):
Transactions Overview | Pro Features Overview
--|--
Screen Shot 2020-01-15 at 12 34 26 | Screen Shot 2020-01-15 at 12 34 18

PRO Feature marker | Enterprise Server Unlock
--|--
Screen Shot 2019-12-06 at 09 54 46 | Screen Shot 2019-12-06 at 09 55 38

Offer Card | -
--|--
Bildschirmfoto 2019-12-12 um 14 47 28 | -

## Pending works
- [x] integrate with Shortcuts app extension
- [x] set up additional IAPs for single features in App Store Connect
- [x] additional testing and polishing

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## QA

Test plan:

https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%201.3.0/Licensing.md

BUGS & IMPROVEMENTS:

- [x] (1) Path in shortcuts message https://github.com/owncloud/ios-app/pull/571#issuecomment-578696897 [FIXED]
- [X] (2) Second account has the features unlocked https://github.com/owncloud/ios-app/pull/571#issuecomment-578709094 [WONT FIX]
- [X] (3) Manage subscriptions ins missing https://github.com/owncloud/ios-app/pull/571#issuecomment-578741449 [WONT FIX]
- [X] (4) 14 days trial is missing https://github.com/owncloud/ios-app/pull/571#issuecomment-578742121
- [x] (5) Features in iOS12 https://github.com/owncloud/ios-app/pull/571#issuecomment-579364587 [FIXED]
- [x] (6) Purchasing with no connection https://github.com/owncloud/ios-app/pull/571#issuecomment-579659272
- [X] (7) Improvements in `Purchases` section https://github.com/owncloud/ios-app/pull/571#issuecomment-579679448
- [X] (8) New appID shows features as unlocked https://github.com/owncloud/ios-app/pull/571#issuecomment-581374089 [FIXED]
- [X] (9) Execution of first shortcut after purchasing fails https://github.com/owncloud/ios-app/pull/571#issuecomment-581797974 [FIXED]

### CODE REVIEW

**Comment by hosy** [comment]
@felix-schwarz should we show in this screen also the items "30 Day Pro Trial " and "Pro Features"?
I am not sure if Apple will reject the IAP, if they are not reachable, without selecting a pro feature. What do you think?

Simulator Screen Shot - iPhone 11 Pro - 2020-01-21 at 09 53 59

File: ownCloud/Client/Actions/Action.swift [hunk 1]
```diff
@@ -315,7 +353,13 @@ class Action : NSObject {
 	}
 
 	func provideAlertAction() -> UIAlertAction? {
-		let alertAction = UIAlertAction(title: self.actionExtension.name, style: actionExtension.category == .destructive ? .destructive : .default, handler: { (_ alertAction) in
+		var name = actionExtension.name
+
+		if !isLicensed {
+			name += " ᴾᴿᴼ" // " 🅿🆁🅾"
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Client/Actions/Action.swift]: Fixed.

File: ownCloud/Client/ClientQueryViewController.swift [hunk 2]
```diff
@@ -519,6 +519,12 @@ class ClientQueryViewController: QueryFileListTableViewController, UIDropInterac
 
 	@objc func plusBarButtonPressed(_ sender: UIBarButtonItem) {
 
+//		let licenseOffer = LicenseOffersViewController(withFeature: "document-scanner")
+//
+//		self.present(licenseOffer, animated: true)
+//
+//		return
+
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Client/ClientQueryViewController.swift]: Yes, it can be removed. Made that change in the latest commit.

File: ownCloud/Licensing/Offers/LicenseOfferView.swift [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+				descriptionLabel.rightAnchor.constraint(equalTo: self.rightAnchor)
+			])
+		}
+
+		if offer.type == .subscription {
+			pricingLabel = UILabel()
+			pricingLabel?.translatesAutoresizingMaskIntoConstraints = false
+			guard let pricingLabel = pricingLabel else { return }
+
+			purchaseButton = LicenseOfferButton(subscribeButtonWithTitle: "Subscribe Now".localized, target: self, action: #selector(takeOffer))
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Offers/LicenseOfferView.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Offers/LicenseOfferView.swift [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+			guard let purchaseButton = purchaseButton else { return }
+
+			var pricingLabelText : String = ""
+			var trialTextLength : Int = 0
+
+			if let trialDuration = offer.trialDuration, offer.state(in: environment) != .expired {
+				pricingLabelText = NSString(format: "Try %@ for free.".localized as NSString, trialDuration.localizedDescription) as String
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Offers/LicenseOfferView.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Offers/LicenseOfferView.swift [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+				pricingLabelText = "\(pricingLabelText)\n"
+				trialTextLength = pricingLabelText.count
+
+				pricingLabelText = pricingLabelText.appendingFormat("Then %@ / %@.".localized, offer.localizedPriceTag, offer.subscriptionTermDuration.localizedDescription)
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Offers/LicenseOfferView.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Offers/LicenseOfferView.swift [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+			} else {
+				// No trial available (either in general, or because user already has subscribed once)
+				pricingLabelText = pricingLabelText.appendingFormat("%@ / %@ – starting immediately".localized, offer.localizedPriceTag, offer.subscriptionTermDuration.localizedDescription)
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Offers/LicenseOfferView.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Offers/LicenseOfferView.swift [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+		switch offer.state(in: environment) {
+			case .uncommitted, .expired:
+				purchaseButton?.isEnabled = true
+
+			case .unavailable:
+				purchaseButton?.isEnabled = false
+
+			case .redundant:
+				purchaseButton?.isEnabled = false
+				purchaseButton?.setTitle("Unlocked".localized, for: .normal)
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Offers/LicenseOfferView.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Offers/LicenseOffersViewController.swift [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	var cardHeaderView : UIView?
+
+	func setupHeaderView() {
+		let headerView = UIView()
+		let titleView = UILabel()
+		let fontSize : CGFloat = 22
+
+		let headerText = NSMutableAttributedString(string: "Pro Features".localized, attributes: [
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Offers/LicenseOffersViewController.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Offers/LicenseOffersViewController.swift [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+			sections.append(subSection)
+		}
+
+		let restoreSection = StaticTableViewSection()
+
+		restoreSection.add(row: StaticTableViewRow(rowWithAction: { (_, _) in
+			OCLicenseManager.shared.restorePurchases(on: self, with: { (_) in
+				self.composeSections()
+			})
+		}, title: "Restore purchases".localized, alignment: .center))
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Offers/LicenseOffersViewController.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Tools/OCLicenseManager+AppStore.swift [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+import UIKit
+import ownCloudApp
+
+extension OCLicenseManager {
+	@objc func restorePurchases(on viewController: UIViewController, with completionHandler: OCLicenseAppStoreRestorePurchasesCompletionHandler? = nil) {
+		if let appStoreProvider = OCLicenseManager.appStoreProvider {
+			let hud : ProgressHUDViewController? = ProgressHUDViewController(on: nil)
+
+			hud?.present(on: viewController, label: "Restoring purchases…".localized)
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Tools/OCLicenseManager+AppStore.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Tools/OCLicenseManager+AppStore.swift [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+			appStoreProvider.restorePurchases(completionHandler: { (error) in
+				let completion = {
+					OnMainThread {
+						if let error = error {
+							let alert = ThemedAlertController(title: "Error restoring purchases".localized, message: error.localizedDescription, preferredStyle: .alert)
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Tools/OCLicenseManager+AppStore.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Transactions/LicenseTransactionsViewController.swift [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+ */
+
+import UIKit
+import ownCloudApp
+
+class LicenseTransactionsViewController: StaticTableViewController {
+	init() {
+		super.init(style: .grouped)
+
+		self.navigationItem.title = "Purchases".localized
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Transactions/LicenseTransactionsViewController.swift]: Fixed in the latest commit.

File: ownCloud/Licensing/Transactions/LicenseTransactionsViewController.swift [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+			UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
+		]
+
+		fetchTransactions()
+	}
+
+	func fetchTransactions() {
+		OCLicenseManager.shared.retrieveAllTransactions(completionHandler: { (error, transactionsByProvider) in
+			if let error = error {
+				let alert = UIAlertController(title: "Error fetching transaction".localized, message: error.localizedDescription, preferredStyle: .alert)
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Licensing/Transactions/LicenseTransactionsViewController.swift]: Fixed in the latest commit.

File: ownCloud/Settings/PurchasesSettingsSection.swift [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+class PurchasesSettingsSection: SettingsSection {
+	// MARK: - More Settings Cells
+
+	private var purchasesRow: StaticTableViewRow?
+	private var transactionsRow: StaticTableViewRow?
+
+	override init(userDefaults: UserDefaults) {
+		super.init(userDefaults: userDefaults)
+
+		self.headerTitle = "In-App Purchases".localized
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloud/Settings/PurchasesSettingsSection.swift]: Fixed in the latest commit.

File: ownCloudAppFramework/Licensing/Offer/OCLicenseDuration.m [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+		case OCLicenseDurationUnitYear:
+			if (_length == 1)
+			{
+				format = OCLocalized(@"year");
+			}
+			else
+			{
+				format = OCLocalized(@"%lu years");
+			}
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloudAppFramework/Licensing/Offer/OCLicenseDuration.m]: Fixed in the latest commit.

File: ownCloudAppFramework/Licensing/Offer/OCLicenseOffer.m [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+#pragma mark - Price information
+- (NSString *)localizedPriceTag
+{
+	if (_localizedPriceTag == nil)
+	{
+		if (_price == nil)
+		{
+			return (OCLocalized(@"Free"));
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloudAppFramework/Licensing/Offer/OCLicenseOffer.m]: Fixed in the latest commit.

File: ownCloudAppFramework/Licensing/Providers/Enterprise/OCLicenseEnterpriseProvider.m [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+@implementation OCLicenseEnterpriseProvider
+
+#pragma mark - Init
+- (instancetype)initWithUnlockedProductIdentifiers:(NSArray<OCLicenseProductIdentifier> *)unlockedProductIdentifiers
+{
+	if ((self = [super initWithIdentifier:OCLicenseProviderIdentifierEnterprise]) != nil)
+	{
+		_unlockedProductIdentifiers = unlockedProductIdentifiers;
+		self.localizedName = OCLocalized(@"Enterprise");
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloudAppFramework/Licensing/Providers/Enterprise/OCLicenseEnterpriseProvider.m]: Fixed in the latest commit.

File: ownCloudAppFramework/Licensing/README.md [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+	- `OCLicenseObserver`
+		- app code
+
+## Examples
+
+#### Registering features and products
+```objc
+// Register features
+[OCLicenseManger.sharedLicenseManager registerFeature:[OCLicenseFeature featureWithIdentifier:@"feature.document-scanning" localizedName:@"Document scanning" localizedDescription:nil]];
+[OCLicenseManger.sharedLicenseManager registerFeature:[OCLicenseFeature featureWithIdentifier:@"feature.push-notifications" localizedName:@"Push notification" localizedDescription:nil]];
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloudAppFramework/Licensing/README.md]: Fixed that an a couple of other errors. Also added a new bit of sample code for simple status checks.

File: ownCloudAppShared/Intent/IntentSettings.swift [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+			environment = OCLicenseEnvironment(bookmark: bookmark)
+		}
+
+		if let environment = environment {
+			return (OCLicenseManager.shared.authorizationStatus(forFeature: .shortcuts, in: environment) == .granted)
+		}
+
+		return false
+	}
+}
```
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloudAppShared/Intent/IntentSettings.swift]: Yeah, I added ClassSettings support in the latest commit.

**Comment by felix-schwarz** [comment]
> @felix-schwarz should we show in this screen also the items "30 Day Pro Trial " and "Pro Features"?
> I am not sure if Apple will reject the IAP, if they are not reachable, without selecting a pro feature. What do you think?

The goal of that screen is not to be a traditional purchase screen - but rather the presentation of a list of Pro Features, their unlock status - and a way to see in which ways they are available for purchase. I.e. tapping "Unlock" next to "Document Scanner" will bring up the actual purchase interface.

In that interface:
- the one-time IAP
- the yearly subscription
- a way to restore purchases (<= as required by Apple's guidelines)

I don't think this should be an issue for App Review. After all, the list of Pro Features is not the actual purchase interface.

File: ownCloud/AppDelegate.swift [hunk 20]
```diff
@@ -80,8 +85,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 		OCExtensionManager.shared.addExtension(LinksAction.actionExtension)
 		OCExtensionManager.shared.addExtension(FavoriteAction.actionExtension)
 		OCExtensionManager.shared.addExtension(UnfavoriteAction.actionExtension)
-		// OCExtensionManager.shared.addExtension(ScanAction.actionExtension)
-
+		OCExtensionManager.shared.addExtension(ScanAction.actionExtension)
```
Comment (hosy) [inline ownCloud/AppDelegate.swift]: `ScanAction` should be only available for >= iOS 13

Comment (felix-schwarz) [reply ownCloud/AppDelegate.swift]: That's already dealt with in the `ScanAction` code, by returning `.none` for` applicablePosition` when not running under iOS 13:

https://github.com/owncloud/ios-app/blob/eb3e5160c014917c1f43cdc11c2b3a8e9a5f29fb/ownCloud/Client/Actions/Scanner/ScanAction.swift#L43

That way, this compatibility/implementation detail of `ScanAction` is fully contained in the action itself and doesn't leak to the `AppDelegate`.

**Comment by jesmrec** [comment]
Let's start with the QA here

**Comment by jesmrec** [comment]
### (1) (wording)

if you try to create a shortcut with no license, this is the message:

Screenshot 2020-01-27 at 11 59 59

Just to be more accurate, the path should be:

Settings > Pro Feafures > Shortcuts

instead of Settings > Premium Features > Shortcuts

**Comment by jesmrec** [comment]
### (2)

Check these steps out:

1. Create a new sandbox user and add it to the device
2. Install the app and add an account -> pro features are locked
3. Add a second account to the device, in a different server

Current: in the second account, pro features are unlocked. 
Expected: both accounts have the features either locked or unlocked (in that specific case must be locked)

iPhoneX 13.3
app commit: `1c520d62` , sdk commit: `c8fdc79`

**Comment by jesmrec** [comment]
### (3)

1. Get the pro subscription package
2. In Settings > Purchases, check the suscription and click on "Manage suscription"

Current: black page
Expected: information about the suscriptions and options to handle (?)

**Comment by jesmrec** [comment]
### (4)

The purchasing card is announcing 14 days of free trial in case of pro suscription, then a one-year suscription to all pro features.

Screenshot 2020-01-27 at 14 16 20

Not sure if it is only in case of sandbox user, but the suscription starts at the moment and finish exactly 1 hour later (for sandbox users, 1 year of suscription is 1 hour till renew). No 14 days period there.

**Comment by jesmrec** [comment]
### (5)

1. Install the app in a iOS12 device

Current: Pro features are available in Settings. 
Expected: Pro features (scanner and shortcuts) are iOS13 features, and should not be available in those devices in which they are not.

File: ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+	func setupLicenseManagement() {
+		if OCLicenseManagerHasBeenSetup {
+			return
+		}
+
+		OCLicenseManagerHasBeenSetup = true
+
+		// Set up features and products
+		let documentScannerFeature = OCLicenseFeature(identifier: .documentScanner, name: "Document Scanner".localized, description: "Scan documents and photos with your camera.".localized)
+		let shortcutsFeature = OCLicenseFeature(identifier: .shortcuts, name: "Shortcuts Actions".localized, description: "Use ownCloud actions in Shortcuts.".localized)
```
Comment (jesmrec) [inline ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift]: Please, do not use the word `ownCloud` in the code (branding issues)

**Comment by jesmrec** [comment]
### (6)

1. Remove connection form device
2. Purchase the scanner (f.ex)

Current: Scanner not purchased, and no feedback about the problem
Expected: Error message (?)

iPhoneX v13.3
app commit: 1c520d62 , sdk commit: c8fdc79

**Comment by jesmrec** [comment]
### (7)

About the Settings > Purchases section

`Purchased App Version` -> this is always 1.0, is it the version of the IAP implementation??
`Receipt Date`-> seems to be the latest purchase/renewal date. Isn't it?

Every purchase/subscription section contains several fields, that are correct. But, it is included a `Quantity` field that is always `1`, but no more than one items can be purchased at the time (at least, for the moment). Did you think in any scenario in which will make sense to purchase more than one item (i guess, any consumable item). If not, i'd get rid of it.

**Comment by felix-schwarz** [comment]
@jesmrec 

(1) Fixed.

(2) Pro Features are unlocked for OC Enterprise servers. Chances are that second server was one running the Enterprise version of OC.

(3) The subscription management UI is brought up by Apple. It is my understanding that this UI shows up blank for the App Store sandbox, but becomes functional once the IAPs are live in the App Store.

(4) It's my understanding that this is indeed sandbox-specific behaviour to better allow testing.

(5) Good point. Since I understand all of the planned Pro Features require iOS 13, it's probably best to not show the IAP interface on older iOS versions. I made that change.

(6) What happens if you bring back the connection after making the purchase?

(7) `Purchased App Version` is the version the user first downloaded the app from the App Store. `Receipt Date` refers to the receipt issued by the App Store describing all purchases and subscriptions. This will typically be the date of the last purchase, subscription or IAP restore.
`Quantity` is a number provided in the App Store receipt. I believe it's number can differ from `1` if Apple's Volume Licensing Program was used to purchase an item.

File: ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift (hunk 21 above)
Comment (felix-schwarz) [reply ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift]: I think in this case it's ok since it describes an IAP - which, as far as I can see - will only ever be available in the unbranded OC app.

**Comment by jesmrec** [comment]
(1) -> fixed

(2) -> you are right... about this one i have to ellaborate a bit more (more test cases to do in the 2nd round)

(3) -> nothing to do with sandbox users then

(4) -> The point is... with real IAPs, will the message be displayed? 

(5) -> iOS12 fixed

(6) -> nothing happens. It seems that tries to do the purchase with no connection, it is not able, and everything finishes. If the connection is recovered, you have to select again the option:

ezgif com-optimize (3)

(7) -> will recheck in the 2nd round

File: ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift (hunk 21 above)
Comment (jesmrec) [reply ownCloudAppShared/Intent/Extensions/OCLicenseManager+Setup.swift]: 👍

**Comment by felix-schwarz** [comment]
> (4) -> The point is... with real IAPs, will the message be displayed?

It definitely should be displayed for these, too, yes. There's a small nuance here, btw, where this message is only shown before the user starts the subscription for the first time. A different message is shown once the user already had a subscription in the past - because trial times are only available once in the App Store.

> (6) -> nothing happens. It seems that tries to do the purchase with no connection, it is not able, and everything finishes. If the connection is recovered, you have to select again the option:

Thanks for the video. I've added error reporting here in the latest commit, so it now brings up an alert.

**Comment by jesmrec** [comment]
### (8)

This one sounds a corner case...

1. Add a sandbox user to test
2. Purchase the IAP features with the user
3. Switch to another test user (Apple ID) in device Settings > iTunes & App Store, which did not purchase before (new user)
4. Restore purchases

Current: In the purchase card, the features appear as unlocked (you see in the background they are locked). 

Screen Shot 2020-02-03 at 12 24 50

Expected: New user did not purchase before, so the features are locked. In the purchase card, the price is shown

iOS13.3 
commit `49f05cab`

**Comment by felix-schwarz** [comment]
@jesmrec Good catch! Commit https://github.com/owncloud/ios-app/pull/571/commits/ceeaaa275ce0fc73900f9e5cb8764e4e83d388b1 should address it.

**Comment by jesmrec** [comment]
### (9)

this is a small glitch

1. Purchase the shortcuts feature
2. Open shortcuts app and create a simple shortcut (f. ex, create a folder)
3. Execute the shortcut

Current: first time you execute the shortcut, the dialog to purchase the pro feature is shown. Second execution works correctly

Expected: once you have purchased the feature, the first (and following) execution runs with no errors regarding the status of the IAP

iPadPro v13.3 
commit  `ceeaaa2`

**Comment by jesmrec** [comment]
Current status

(1), (2), (3), (5), (6) -> DONE

(4) -> this is for Apple Store purchases, showed once till the feature is purchased. Not much to do with it right now. 

(7) -> the `Receipt Date` is correct. The fact `Purchased App Version` shows 1.0 is because the app is installed from XCode and not downloaded from App Store. Isn't it?

(8) -> i still reproduce the problem with commit `49f05cab`

(9) -> just reported

**Comment by felix-schwarz** [comment]
@jesmrec 

> (7) -> the `Receipt Date` is correct. The fact `Purchased App Version` shows 1.0 is because the app is installed from XCode and not downloaded from App Store. Isn't it?

No, that's always the first version of the app that you installed from the App Store.

For someone downloading the app for the first time today, it'll be version 1.2.1.

> (8) -> i still reproduce the problem with commit `49f05cab`

This remains a head-scratcher for me, and I'll have to do more testing to hopefully catch it.

> (9) -> just reported

Fixed in https://github.com/owncloud/ios-app/pull/571/commits/a4457f4b7cb26236cb46dc83705f5fb125d820db . That was a timing issue with a fairly small corridor (0.01 seconds on my iPhone 6s – I had to try a few times to reproduce it). Good catch!

**Comment by felix-schwarz** [comment]
> > (8) -> i still reproduce the problem with commit `49f05cab`
> 
> This remains a head-scratcher for me, and I'll have to do more testing to hopefully catch it.

I tried multiple times, but couldn't reproduce it. When switching sandbox accounts in Settings, iOS doesn't seem to switch the account for the ownCloud app here. Even if I restore purchases then (multiple times), the app won't get a new or different App Store receipt.

**Comment by jesmrec** [comment]
(7) -> thanks for clarification

(9) -> fixed

(8) -> about your comment

> I tried multiple times, but couldn't reproduce it. When switching sandbox accounts in Settings, iOS doesn't seem to switch the account for the ownCloud app here. Even if I restore purchases then (multiple times), the app won't get a new or different App Store receipt.

What i expect is that a new Apple ID that contains a different purchase history, resets the history of the previous Apple ID (am i wrong?)

This is the behaviour i see with the commit `f4bb1af5`. This is slightly different as the one i defined before:

I have two AppleIDs:  one that have both features purchased (1) , and another one with no purchases (2).

1. Add (1) in device settings. 
2. In app, restore purchases: both features are unlocked
3. Switch to (2) in device Settings
4. In app, restore purchases

Current:

- Features keep unlocked
- List of purchases is not cleared

By removing the app and installing again with (2), the list of purchases is clear after restoring purchases.

I can send you the sandbox users i reproduce the problem with, so you can test directly with them. Also, i can record a video, get logs... whatever is needed.

**Comment by felix-schwarz** [comment]
Re (8): I think this "works as designed by Apple" then:
- you can install apps purchased through *different* accounts on the same iOS device and use them alongside
- the only way that can technically work is that iOS apps
    - remain tied to the Apple ID they were installed with
    - as part of this keep their App Store receipt with all in-app purchases from the Apple ID the app was installed with

**Comment by michaelstingl** [comment]
@felix-schwarz So, nothing left to do? Merge?

**Comment by felix-schwarz** [comment]
> @felix-schwarz So, nothing left to do? Merge?

Yes, if @jesmrec doesn't see any remaining issues, either, I'd vote to merge the PR, too.
